### PR TITLE
Events Map JPA implementation

### LINKED ISSUE #9667: JPA map storage: Events no-downtime store
### Description

Create a JPA map storage implementation for Events area

### PR DESCRIPTION
**Pull request by sguilhen**
Closes #9667

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaModelCriteriaBuilder.java
Comment (sguilhen) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaModelCriteriaBuilder.java:137]: @vramik I've borrowed the above code from your authz PR. Once that is merged I'll update.

File: model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml
Comment (sguilhen) [inline model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml:45]: Question: should the expiration be part of this index? It is used on every query to avoid returning expired entities.

File: model/map-jpa/src/main/resources/META-INF/auth-events/jpa-auth-events-changelog-1.xml
Comment (sguilhen) [inline model/map-jpa/src/main/resources/META-INF/auth-events/jpa-auth-events-changelog-1.xml:60]: Question: This is similar to the attributes tables we have for other entities, and those have an index for the name/value. I haven´t added such index here because the details are never queried by name or value, they are always loaded in their entirety, so to me only the fk_root index should be enough

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/event/admin/delegate/JpaAdminEventDelegateProvider.java
Comment (sguilhen) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/event/admin/delegate/JpaAdminEventDelegateProvider.java:67]: @vramik I've added the delegate for now, although we might decide to simply load everything - including the metadata - in read with parameters. You mentioned you would follow this approach for the user sessions, and if you do I'll update this one to be consistent with what you are doing there.

File: model/map/src/main/java/org/keycloak/models/map/events/MapAdminEventQuery.java
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/events/MapAdminEventQuery.java:93]: I believe we shouldn't transform `ResouceType` to `String` at this level, the searchable field states that it expects `ResourceType.class` as a filed type. I'd suggest rather change `MapFiledsPredicates` class as mentioned in https://github.com/keycloak/keycloak/pull/12299#discussion_r888707679 but fist we should decide how the search should behave for ResourceType.CUSTOM (https://github.com/keycloak/keycloak/pull/12299#discussion_r889222201)

File: model/map/src/main/java/org/keycloak/models/map/events/MapAuthEventQuery.java
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/events/MapAuthEventQuery.java:55]: out of curiosity: why this change?

File: model/map-jpa/src/main/resources/META-INF/auth-events/jpa-auth-events-changelog-1.xml
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/auth-events/jpa-auth-events-changelog-1.xml:64]: +1 for only having `fk_root` index

File: model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml:45]: Not sure to be honest. It might make sense to have it due to being part of each search as you said. On the other hand expiration field would be changed very often which would mean the need for re-calculating the index very often. 

I'd go without it for now as database should be able to uset currently existing index and then "just" filter non-expired objects. And if it turns out the index is needed it can be added later, wdyt?

**Review by vramik — COMMENTED** [review]
Thank you @sguilhen for the PR! I've done first round of review and have some comments, please see inline. I'll take another look once https://github.com/keycloak/keycloak/pull/12299 is merged.

File: model/map/src/main/java/org/keycloak/models/map/events/MapAdminEventQuery.java
Comment (sguilhen) [reply model/map/src/main/java/org/keycloak/models/map/events/MapAdminEventQuery.java:93]: Agreed, will revert this change.

File: model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml
Comment (sguilhen) [reply model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml:45]: Seems reasonable to me, will remove it from the index.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/event/admin/delegate/JpaAdminEventDelegateProvider.java
Comment (sguilhen) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/event/admin/delegate/JpaAdminEventDelegateProvider.java]: Changed, both event entities no longer have a delegate.

File: testsuite/integration-arquillian/tests/base/pom.xml
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/pom.xml:890]: Could you please use a different separator than `.` for `eventsStore.admin`, e.g. `eventsStore-admin` or (perhaps better) even `adminEventsStore`?

File: model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml
Comment (sguilhen) [reply model/map-jpa/src/main/resources/META-INF/admin-events/jpa-admin-events-changelog-1.xml]: Nothing really. I was experimenting with different indexes and this ended up in the PR by mistake. Removed.

**Review by vramik — DISMISSED** [review]
LGTM, thank you @sguilhen! 

Depending whether https://github.com/keycloak/keycloak/pull/12230 will be merged first, this one will require rebase due to the `EXPIRATION` removal.

**Comment by hmlnarik** [comment]
@sguilhen  #12230 is merged, could you please rebase?

**Comment by sguilhen** [comment]
@hmlnarik Rebased!

**Review by vramik — APPROVED** [review]
Thanks @sguilhen, great work!

**Review by vramik — COMMENTED** [review]
@sguilhen one question popped up - I assumed we should filter expired entities on the database level, but I cannot see where it is done. I can see expired entities are handled on model level within `MapEventStoreProvider`, but that way we'd load all entities from database. Did I got it wrong?

**Comment by vramik** [comment]
> @sguilhen one question popped up - I assumed we should filter expired entities on the database level, but I cannot see where it is done. I can see expired entities are handled on model level within `MapEventStoreProvider`, but that way we'd load all entities from database. Did I got it wrong?

I've just seen https://github.com/keycloak/keycloak/issues/12623. Just to be sure, we'd tackle this within this issue?

**Review by hmlnarik — APPROVED** [review]
Approving per @vramik 's review

**Pull request merged**
